interface/ceed-basis.c

```diff
@@ -829,6 +829,161 @@ int CeedBasisCreateHdiv(Ceed ceed, CeedElemTopology topo, CeedInt num_comp,
+
+  @param[in] basis_fine      Fine CeedBasis to prolong to
+  @param[in] basis_coarse    Coarse CeedBasis to prolong from
+  @param[out] basis_prolong  Address of the variable where the newly created
+                               CeedBasis will be stored.
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedBasisCreateProlongation(CeedBasis basis_fine, CeedBasis basis_coarse,
+                                CeedBasis *basis_prolong) {
+  int ierr;
+  Ceed ceed;
+  ierr = CeedBasisGetCeed(basis_fine, &ceed); CeedChk(ierr);
+
+  // Create prolongation matrix
+  CeedScalar *interp_c_to_f;
+  ierr = CeedBasisCreateProlongationMatrix(basis_fine, basis_coarse,
+         &interp_c_to_f); CeedChk(ierr);
+
+  // Build basis  // Build prolongation matrix
+  bool is_tensor;
+  CeedInt dim, num_comp;
+  CeedScalar *q_ref, *q_weight, *grad;
+  ierr = CeedBasisIsTensor(basis_fine, &is_tensor); CeedChk(ierr);
+  ierr = CeedBasisGetDimension(basis_fine, &dim); CeedChk(ierr);
+  ierr = CeedBasisGetNumComponents(basis_fine, &num_comp); CeedChk(ierr);
+  if (is_tensor) {
+    CeedInt P_1d_f, P_1d_c;
+    ierr = CeedBasisGetNumNodes1D(basis_coarse, &P_1d_c); CeedChk(ierr);
+    ierr = CeedBasisGetNumNodes1D(basis_fine, &P_1d_f); CeedChk(ierr);
+    ierr = CeedCalloc(P_1d_f, &q_ref); CeedChk(ierr);
+    ierr = CeedCalloc(P_1d_f, &q_weight); CeedChk(ierr);
+    ierr = CeedCalloc(P_1d_f*P_1d_c*dim, &grad); CeedChk(ierr);
+    ierr = CeedBasisCreateTensorH1(ceed, dim, num_comp, P_1d_c, P_1d_f,
+                                   interp_c_to_f, grad, q_ref, q_weight, basis_prolong);
+    CeedChk(ierr);
+  } else {
+    CeedElemTopology topo;
+    ierr = CeedBasisGetTopology(basis_fine, &topo); CeedChk(ierr);
+    CeedInt num_nodes_c, num_nodes_f;
+    ierr = CeedBasisGetNumNodes(basis_coarse, &num_nodes_c); CeedChk(ierr);
+    ierr = CeedBasisGetNumNodes(basis_fine, &num_nodes_f); CeedChk(ierr);
+    ierr = CeedCalloc(num_nodes_f*dim, &q_ref); CeedChk(ierr);
+    ierr = CeedCalloc(num_nodes_f, &q_weight); CeedChk(ierr);
+    ierr = CeedCalloc(num_nodes_f*num_nodes_c*dim, &grad); CeedChk(ierr);
+    ierr = CeedBasisCreateH1(ceed, topo, num_comp, num_nodes_c, num_nodes_f,
+                             interp_c_to_f, grad, q_ref, q_weight, basis_prolong);
+    CeedChk(ierr);
+  }
+
+  // Cleanup
+  ierr = CeedFree(&interp_c_to_f); CeedChk(ierr);
+  ierr = CeedFree(&q_ref); CeedChk(ierr);
+  ierr = CeedFree(&q_weight); CeedChk(ierr);
+  ierr = CeedFree(&grad); CeedChk(ierr);
+
+  return CEED_ERROR_SUCCESS;
+}
+
+/**
+  @brief Create the interpolation matrix for prolongation from the nodes of `basis_c`
+           to the nodes of `basis_f`.
+
+  @param[in] basis_fine      Fine CeedBasis to prolong to
+  @param[in] basis_coarse    Coarse CeedBasis to prolong from
+  @param[out] interp_c_to_f  Address of the variable where the newly created
+                               prolongation matrix will be stored.
```

Comment:
Let's just document (in the man page, in addition to the check you have below) that these must have matching quadratures.

And perhaps write this creates a $B_p$ such that $B_f B_p = B_c$. This is possible any time it's actually a prolongation, but there may be times when one wants to reverse the roles, asking for $B_c B_p \approx B_f$. There are stabilization methods that use this sort of operator so we could call this something like `CeedBasisCreateProjectionMatrix` so that it makes sense for both uses.

Reply:
Good call on all counts.